src/librustc/traits/error_reporting.rs

```diff
@@ -347,6 +347,53 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         }
     }
 
+    fn describe_generator(&self, body_id: hir::BodyId) -> Option<&'static str> {
+        self.tcx.hir().body(body_id).generator_kind.map(|gen_kind| {
+            match gen_kind {
+                hir::GeneratorKind::Gen => { "a generator" },
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Block) => { "an async block" },
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Fn) => { "an async function" },
+                hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Closure)
+                    => { "an async closure" },
+            }
+        })
+    }
+
+    /// Used to set on_unimplemented's `ItemContext`
+    /// to be the enclosing (async) block/function/closure
+    fn describe_enclosure(&self, hir_id: hir::HirId) -> Option<&'static str> {
+        let hir = &self.tcx.hir();
+        let node : Node<'_> = hir.get(hir_id);
```

Comment:
```suggestion
        let node : Node<'_> = hir.find(hir_id)?;
```

Reply:
You can use ? returns with options as well as results?? That's a thing of beauty.